src/LEPTON/fix_epot_lepton.cpp

```diff
@@ -0,0 +1,324 @@
+    for (size_t i = 0; i < 3; i++) {
+      try { 
+        (*cexpr).getVariableReference(variableNames[i]);
+      }
+      catch (Lepton::Exception &) {
+        has_ref.back()[i] = false;
+      }
+    }
+  }
+  // virial setup
+  v_init(vflag);
+
+  // update region if necessary
+  if (region) region->prematch();
+
+  // fsum[0] = "potential energy" for added force
+  // fsum[123] = extra force added to atoms
+  fsum[0] = fsum[1] = fsum[2] = fsum[3] = 0.0;
+  force_flag = 0;
+
+  double fx, fy, fz;
+  double v[6], unwrap[3];
+
+  for (int i = 0; i < nlocal; i++) {
+    if (mask[i] & groupbit) {
+      if (region && !region->match(x[i][0], x[i][1], x[i][2])) continue;
+      domain->unmap(x[i], image[i], unwrap);
+      
+      // substitute x, y, z if they exist           
+      for (size_t j = 0; j< cexprs.size(); j++) {
+        if (has_ref[j][0]) (*cexprs[j]).getVariableReference("x") = unwrap[0];
+        if (has_ref[j][1]) (*cexprs[j]).getVariableReference("y") = unwrap[1];
+        if (has_ref[j][2]) (*cexprs[j]).getVariableReference("z") = unwrap[2];
+      }
+
+      // charges
+      // force = q E, potential energy = q phi
+      if (atom->q_flag) {
+        fx = -qe2f * q[i] * dphi_x.evaluate();
+        fy = -qe2f * q[i] * dphi_y.evaluate();
+        fz = -qe2f * q[i] * dphi_z.evaluate();
+        f[i][0] += fx;
+        f[i][1] += fy;
+        f[i][2] += fz;
+
+        fsum[0] += qqr2e * q[i] * phi.evaluate();
+        fsum[1] += fx;
+        fsum[2] += fy;
+        fsum[3] += fz;
+        if (evflag) {
+          v[0] = fx * unwrap[0];
+          v[1] = fy * unwrap[1];
+          v[2] = fz * unwrap[2];
+          v[3] = fx * unwrap[1];
+          v[4] = fx * unwrap[2];
+          v[5] = fy * unwrap[2];
+          v_tally(i, v);
+        }
+      }
+
+      // dipoles
+      // force = (mu dot D) E, torque = mu cross E, potential energy = - mu dot E
+      if (atom->mu_flag) {
+        double tx, ty, tz;
+        auto ex = -dphi_x.evaluate();
+        auto ey = -dphi_y.evaluate();
+        auto ez = -dphi_z.evaluate();
+        tx = ez * mu[i][1] - ey * mu[i][2];
+        ty = ex * mu[i][2] - ez * mu[i][0];
+        tz = ey * mu[i][0] - ex * mu[i][1];
+        t[i][0] += qqr2e * tx;
+        t[i][1] += qqr2e * ty;
+        t[i][2] += qqr2e * tz;
+        fsum[0] -= qqr2e * (mu[i][0] * ex + mu[i][1] * ey + mu[i][2] * ez);
+
+        fx = -qe2f *
+            (mu[i][0] * dphi_xx.evaluate() + mu[i][1] * dphi_xy.evaluate() +
+             mu[i][2] * dphi_xz.evaluate());
+        fy = -qe2f *
+            (mu[i][0] * dphi_xy.evaluate() + mu[i][1] * dphi_yy.evaluate() +
+             mu[i][2] * dphi_yz.evaluate());
+        fz = -qe2f *
+            (mu[i][0] * dphi_xz.evaluate() + mu[i][1] * dphi_yz.evaluate() +
+             mu[i][2] * dphi_zz.evaluate());
+
+        f[i][0] += fx;
+        f[i][1] += fy;
+        f[i][2] += fz;
+
+        fsum[1] += fx;
+        fsum[2] += fy;
+        fsum[3] += fz;
+      }
```

Comment:
This fix doesn't do `evtally` for dipoles? It's worth documenting in code comments if there's a reason

Reply:
This is a good catch. I based my code on `fix_efield.cpp` where the `v_tally` is only done for charges under a constant E-field. Since the dipoles will only experience a force under an in-homogeneous field, I'm not sure if that should be included in the Virial.

Reply:
Please note that LAMMPS has fix numdiff and fix numdiff/virial to validate forces and virials from potential energy through finite differences.

Reply:
Thanks, @akohlmey. I looked at the source code and it looks like they only compute the change in energy wrt atomic displacements. In this case, the force on the dipoles should be added to the virial. However, there would still be a missing rotational energy since the dipoles should have atom style sphere for this fix and `fix efield`.